pam_ssh_auth_info.8
.\" Copyright © 2021 - 2025 Eero Häkkinen <Eero+pam-ssh-auth-info@Häkkinen.fi>
.\"
.\" This manual page is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This manual page is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this manual page.  If not, see <http://www.gnu.org/licenses/>.
.if '\*[.T]'html' \{\
.HEAD "<link href=""groff.css"" rel=""stylesheet"" type=""text/css"" />"
.HEAD "<meta name=""viewport"" content=""width=device-width, initial-scale=1.0"" />"
.\}
.TH "pam_ssh_auth_info" "8" "2025-04-21"
.if '\*[.T]'html' .if d HTML-NS \{\
.\" Work-around bug #61915: grohtml: .EX/.EE is not monospaced
.\"             https://savannah.gnu.org/bugs/?61915
.rn EX EX0
.de EX
.	EX0
.	ft C
.	HTML <!--
.	HTML-NS -->
..
.rn EE EE0
.de EE
.	ft
.	EE0
..
.\}

.SH "NAME"
pam_ssh_auth_info \- test SSH authentication information

.SH "SYNOPSIS"
.B  pam_ssh_auth_info.so
.RI [ flag ...]
.RI [ pattern [= pattern ]...]...

.SH "DESCRIPTION"
The pam_ssh_auth_info.so PAM module is designed
to succeed or fail authentication
based on SSH authentication information
consisting of a list of
successfully completed authentication methods and
public credentials (e.g. keys)
used to authenticate the user.
One use is to select whether to load other modules based on this test.

Normally,
the module should be given zero or more patterns as module arguments, and
authentication will succeed
if SSH authentication information matches all of the patterns
(or any of or none of the patterns depending on the options).
If there are no previous successfully completed authentication methods,
SSH authentication information
is not available but missing and
authentication will neither succeed nor fail
(the module will return \fBPAM_IGNORE\fP).

.SH "OPTIONS"
The following \fIflag\fPs are supported:
.TP
.B all_of
All of the \fIpattern\fPs must match.
If zero \fIpattern\fPs are given as module arguments,
authentication will succeed
if SSH authentication information is available.
This is the default.
.TP
.B any_of
At least one of the \fIpattern\fPs must match.
If zero \fIpattern\fPs are given as module arguments,
authentication will fail
if SSH authentication information is available.
.TP
.B debug
Log debugging messages to syslog.
.TP
.BI disable= service \fR[\fP: service \fR[...]]
Disable pattern matching for the services
listed in the colon separator service list.
.TP
.BI enable= service \fR[\fP: service \fR[...]]
Enable pattern matching only for the services
listed in the colon separator service list.
.TP
.B none_of
None of the \fIpattern\fPs may match.
If zero \fIpattern\fPs are given as module arguments,
authentication will succeed
if SSH authentication information is available.
.TP
.B quiet
Do not log failure or success messages to syslog.
.TP
.B quiet_fail
Do not log failure messages to syslog.
.TP
.B quiet_success
Do not log success messages to syslog.
.TP
.BI recursion_limit= limit
Change the recursion limit.
This affects extended patterns and \fB*\fP wildcard patterns.

.SS "PATTERNS"
Any character byte that appears in a pattern,
other than
the extended patterns and
the special pattern characters
described below,
matches itself.

.PP
The special pattern characters have the following meanings:
.TP
.B *
Matches any number of (including zero) word character bytes
but not a word separator (space).
.TP
.B =
Matches an equal-sign (\fB=\fP) or a word separator (space).
This
enables more natural looking patterns and
allows patterns to be written without spaces
avoiding the need for square bracket quoting in PAM configuration files.
.TP
.B ?
Matches any word character byte
but not a word separator (space).
.TP
.B [
A pair of brackets introduces a character byte class.
A character byte class (\fB[\fP...\fB]\fP)
matches any word character byte in the class.
A complemented character byte class (\fB[!\fP...\fB]\fP)
matches any word character byte not in the class.
Neither matches a word separator (space).
A class may contain
character bytes (\fB[abcde]\fP) and
character byte ranges (\fB[a-e]\fP).
.TP
.B \\\\
Preserves the literal meaning of the following character byte.

.PP
The extended patterns have the following meanings:
.TP
.BI "?(" pattern "|" pattern "|" \fR... ")"
Matches zero or one occurence of the given patterns within a word.
Does not match a word separator (space).
.TP
.BI "*(" pattern "|" pattern "|" \fR... ")"
Matches zero or more occurences of the given patterns within a word.
Does not match a word separator (space).
.TP
.BI "@(" pattern "|" pattern "|" \fR... ")"
Matches one occurence of the given patterns within a word.
Does not match a word separator (space).
.TP
.BI "+(" pattern "|" pattern "|" \fR... ")"
Matches one or more occurences of the given patterns within a word.
Does not match a word separator (space).
.TP
.BI "!(" pattern "|" pattern "|" \fR... ")"
Matches anything within a word except one occurence of the given patterns.
Does not match a word separator (space).

.PP
SSH authentication information consists of lines having a format like
.IP
.\" method [\fB/ submethod "][ " auth-method-key       "][ " auth-method-info ]
.IR method [\fB/ submethod "][ " key-type " " key-data "][ " info-word     ]...
.PP
SSH authentication information matches a pattern
if on any such line all the words or the initial words
match the pattern.

.SH "MODULE TYPES PROVIDED"
All module types
(\fBaccount\fP, \fBauth\fP, \fBpassword\fP and \fBsession\fP)
are provided.
That being said,
the \fBauth\fP type is still undoubtedly the most useful one.

.SH "RETURN VALUES"
.TP
.B PAM_AUTH_ERR
Pattern requirements are not met.
SSH authentication information
does not match all of or any of the patterns
(see the \fBall_of\fP and the \fBany_of\fP options) or
matches some of the patterns
(see the \fBnone_of\fP option).
.TP
.B PAM_IGNORE
The pattern matching is
disabled for the service (see the \fBdisable\fP option) or
not enabled for the service (see the \fBenable\fP option) or
SSH authentication information is missing.
.TP
.B PAM_SUCCESS
Pattern requirements are met.
SSH authentication information
matches all of or any of the patterns
(see the \fBall_of\fP and the \fBany_of\fP options) or
matches none of the patterns
(see the \fBnone_of\fP option).

.SH EXAMPLES

.PP
Always load an OATH module rule
(to request and check a one-time password)
but load a password module rule
(to request and check the user password)
only
if there are
no previous successfully completed authentications
(in which case SSH authentication information is missing):
.IP
.EX
auth  requisite         oath-module.so option...
auth  [success=1        \\
       default=ignore]  pam_ssh_auth_info.so
auth  required          password-module.so try_first_pass option...
.EE

.PP
Load an OATH module rule
(to request and check a one-time password)
only
if there are
no previous successfully completed
FIDO authenticator algorithm based
public key authentications
(in which case SSH authentication information either
is missing or
does not match the pattern)
but always load a password module rule
(to request and check the user password):
.IP
.EX
auth  [success=1        \\
       ignore=ignore    \\
       auth_err=ignore  \\
       default=die]     pam_ssh_auth_info.so quiet \\
                            publickey=*sk-*@openssh.com
auth  requisite         oath-module.so option...
auth  required          password-module.so try_first_pass option...
.EE

.PP
Require that there is
at least one previous successfully completed
FIDO authenticator algorithm based
public key authentication
(in which case
the key type
contains \(lqsk-\(rq and
ends with \(lq@openssh.com\(rq).
If multiple public key authentications are required,
only one of them is required to be a FIDO authenticator algorithm based one.
.IP
.EX
auth  requisite  pam_ssh_auth_info.so quiet \\
                     publickey=*sk-*@openssh.com
.EE

.PP
Require that there is
at least one previous successfully completed
FIDO authenticator algorithm based
public key authentication
(in which case
the key type
contains \(lqsk-\(rq and
ends with \(lq@openssh.com\(rq) and
at least one previous successfully completed
non-FIDO
public key authentication
(in which case
the key type
does not contain \(lqsk-\(rq or
does not end with \(lq@openssh.com\(rq).
.IP
.EX
auth  requisite  pam_ssh_auth_info.so quiet \\
                     publickey=*sk-*@openssh.com \\
                     publickey=!(*sk-*@openssh.com)
.EE

.SH "ENVIRONMENT"
.TP
.B SSH_AUTH_INFO_0
SSH authentication information
consisting of a list of
successfully completed authentication methods and
public credentials (e.g. keys)
used to authenticate the user.
This environment variable
is provided by OpenSSH server since version 7.6p1 and
is visible
to PAM auth modules
(beginning with OpenSSH version 7.8p1
if keyboard-interactive challenge-response authentication is used and
beginning with OpenSSH version 9.8p1
if password authentication is used) and
to PAM account, session and password modules
(beginning with OpenSSH version 7.6p1)
but not to user sessions.
.TP
.B SSH_USER_AUTH
A location of a file
containing
SSH authentication information
consisting of a list of
successfully completed authentication methods and
public credentials (e.g. keys)
used to authenticate the user.
This environment variable
is provided by OpenSSH server since version 7.6p1
if the \fBExposeAuthInfo\fP server option is enabled and
is visible
to user sessions
but not to PAM modules.
This may be a useful source
for creating proper patterns.

.SH "NOTES"
SSH authentication information is available
to PAM auth modules
beginning with OpenSSH version 7.8p1
if keyboard-interactive challenge-response authentication is used and
beginning with OpenSSH version 9.8p1
if password authentication is used.
SSH authentication information is available
to PAM account, session and password modules
beginning with OpenSSH version 7.6p1.
.PP
In order to make this module useful,
the following \fBsshd_config\fP(5) options should be set
(either explicitly or via defaults):
.TP
.BR \%AuthenticationMethods
Should contain comma-separated lists of authentication method names
with
\fBkeyboard-interactive\fP,
\fBkeyboard-interactive:pam\fP or
\fBpassword\fP
as the last list items
so that there are
successfully completed authentication methods and
public credentials (e.g. keys)
during the PAM authentication.
For example
.RS
.IP
.EX
.BR AuthenticationMethods " " publickey,keyboard-interactive
.EE
.PP
for
mandatory public key authentication and
mandatory PAM authentication
using keyboard-interactive challenge-response authentication or
.IP
.EX
.BR AuthenticationMethods " " publickey,keyboard-interactive " " keyboard-interactive
.EE
.PP
for
optional public key authentication and
mandatory PAM authentication
using keyboard-interactive challenge-response authentication.
.RE
.TP
.BR \%ChallengeResponseAuthentication ", " \%KbdInteractiveAuthentication ", " \%PasswordAuthentication ", etc."
The relevant ones should be set to \fByes\fP.
.TP
.BR \%UsePAM
Should be set to \fByes\fP.

.SH "SEE ALSO"
.BR \%pam (7),
.BR \%sshd_config (5)

.na
.UR https://github.Eero.xn--Hkkinen-5wa.fi/pam-ssh-auth-info/
Home page for pam_ssh_auth_info
.UE
.ad

.SH "AUTHOR"
.na
Eero Häkkinen <Eero+pam-ssh-auth-info@Häkkinen.fi>
.ad

.SH "COPYRIGHT"
.na
Copyright © 2021 - 2025 Eero Häkkinen <Eero+pam-ssh-auth-info@Häkkinen.fi>
.ad

This manual page is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This manual page is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this manual page.  If not, see <http://www.gnu.org/licenses/>.


debian/changelog
pam-ssh-auth-info (2.0.~~yyyymmdd) UNRELEASED; urgency=low

  * Avoided unneeded measurements of extended patterns.
  * Measured !(...) extended patterns.
  * Improved handling of question marks after asterisks.
  * Fixed handling of *(...) extended patterns after asterisks.
  * Optimized extended patterns.
  * Reduced recursion.
  * Improved handling of character bytes after asterisks.
  * Added more test cases.
  * Improved manual page source format.
  * Clarified documentation of extended patterns.
  * Added debian/*.install and debian/not-install files.
  * Declared an optional build-dependency metapackage.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Mon, 21 Apr 2025 18:44:59 +0300

pam-ssh-auth-info (1.8.20230906) unstable; urgency=low

  * Fixed and updated documentation.
  * Avoided some compiler warnings.
  * Improved tail call optiomization.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Wed, 06 Sep 2023 21:19:19 +0300

pam-ssh-auth-info (1.6.20220918) unstable; urgency=medium

  * Changed license to GPL/LGPL-3+.
  * Changed pam_sm_setcred to succeed or fail based on SSH
    authentication information (instead of merely returning PAM_IGNORE
    unconditionally).
  * Fixed crash on empty extended patterns.
  * Preferred all_of/any_of options over all/any options.
  * Added support for the none_of option.
  * Improved documentation.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Sun, 18 Sep 2022 20:03:04 +0300

pam-ssh-auth-info (1.4.20220116) unstable; urgency=low

  * Added recursion limit.
  * Fixed documentation.
  * Described SSH_USER_AUTH.
  * Implemented tail call optimization for * wildcard patterns.
  * Implemented tail call optimization for extended patterns.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Sun, 16 Jan 2022 16:12:38 +0200

pam-ssh-auth-info (1.2.20211208) unstable; urgency=low

  * Added a missing libpam-dev build dependency (closes: GH-2).
    Thanks to Cyril B. (@cbay).

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Wed, 08 Dec 2021 13:32:50 +0200

pam-ssh-auth-info (1.2.20211201) unstable; urgency=low

  * Fixed handling of unpaired backslashes.
  * Added support for patterns containing extended patterns.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Wed, 01 Dec 2021 18:08:31 +0300

pam-ssh-auth-info (1.0.20210520) unstable; urgency=low

  * Fixed license boilerplates.
  * Renamed license files.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Thu, 20 May 2021 21:42:36 +0300

pam-ssh-auth-info (1.0.20210430) unstable; urgency=low

  * Defined the remaining PAM_SM_ macros.
  * Fixed handling of empty service names.
  * Improved configuration documentation (closes: GH-1).
  * Installed README.md to the doc directory.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Fri, 30 Apr 2021 17:05:55 +0300

pam-ssh-auth-info (1.0.20210223) unstable; urgency=low

  * Added support for patterns containing character byte classes.
  * Added support for patterns containing backslashes.
  * Added version match test.
  * Enabled all hardening features.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Tue, 23 Feb 2021 22:01:17 +0200

pam-ssh-auth-info (1.0.20210124) unstable; urgency=low

  * Provided all module types.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Sun, 24 Jan 2021 23:42:00 +0200

pam-ssh-auth-info (1.0.20210121) unstable; urgency=low

  * Described SSH authentication information.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Thu, 21 Jan 2021 21:27:05 +0200

pam-ssh-auth-info (1.0.20210117) unstable; urgency=low

  * Initial release.

 -- Eero Häkkinen <Eero+pam-ssh-auth-info@xn--Hkkinen-5wa.fi>  Sun, 17 Jan 2021 00:41:51 +0200


pam_ssh_auth_info.c
/*
 * Copyright © 2021 - 2024 Eero Häkkinen <Eero+pam-ssh-auth-info@Häkkinen.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
#	include "config.h"
#endif

#if !defined(HAVE_PAM_SYSLOG) && defined(HAVE_VSYSLOG) && !defined(_DEFAULT_SOURCE)
#	define _DEFAULT_SOURCE
#endif

#undef  NDEBUG
#define NDEBUG
#define PAM_SM_ACCOUNT
#define PAM_SM_AUTH
#define PAM_SM_SESSION
#define PAM_SM_PASSWORD

#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#include <syslog.h>

#if defined(HAVE_SECURITY_PAM_APPL_H) || !defined(PACKAGE_NAME)
#	include <security/pam_appl.h>
#endif
#if defined(HAVE_SECURITY_PAM_EXT_H) || !defined(PACKAGE_NAME)
#	include <security/pam_ext.h>
#endif
#if defined(HAVE_SECURITY_PAM_MODULES_H) || !defined(PACKAGE_NAME)
#	include <security/pam_modules.h>
#endif

#include "line_tokens_match.h"
#include "pam_syslog.h"

/* Check if a string is in a list separated by separators.
 */
static bool
in_list(char const *list, char sep, char const *s) {
	size_t const n = strlen(s);
	for (char const *p = strstr(list, s); p && *p; p = strstr(p + 1, s)) {
		/* Check that the string is either the first item in the list
		 * or is preceded by a separator and that the string is either
		 * the last item in the list or is followed by a separator (so
		 * a mere substring within another list item is not enough).
		 */
		if ((p == list || p[-1] == sep) && (!p[n] || p[n] == sep))
			return true;
	}
	return false;
}

/* Locate the beginning of the next line or the NUL byte.
 */
static char const *
next_line(char const *s) {
	s += strcspn(s, "\n");
	if (*s == '\n')
		++s;
	return s;
}

int
pam_sm_authenticate(
	pam_handle_t *pamh,
	int flags,
	int argc,
	char const **argv
	) {
	(void)flags;
	/* Parse options.
	 */
	bool debug = false;
	char const *disable = NULL;
	char const *enable = NULL;
	enum {
		MATCH_ALL_OF,
		MATCH_ANY_OF,
		MATCH_NONE_OF
	} match_style = MATCH_ALL_OF;
	bool quiet_fail = false;
	bool quiet_success = false;
	unsigned recursion_limit = 100u;
	for (; argc > 0; --argc, ++argv) {
		if (strncmp(*argv, "all", 3) == 0 && (
			strcmp(*argv + 3, "_of") == 0 ||
			strcmp(*argv + 3, "") == 0
			))
			match_style = MATCH_ALL_OF;
		else if (strncmp(*argv, "any", 3) == 0 && (
			strcmp(*argv + 3, "_of") == 0 ||
			strcmp(*argv + 3, "") == 0
			))
			match_style = MATCH_ANY_OF;
		else if (strcmp(*argv, "debug") == 0)
			debug = true;
		else if (strncmp(*argv, "disable=", 8) == 0)
			disable = *argv + 8;
		else if (strncmp(*argv, "enable=", 7) == 0)
			enable = *argv + 7;
		else if (strcmp(*argv, "none_of") == 0)
			match_style = MATCH_NONE_OF;
		else if (strcmp(*argv, "quiet") == 0)
			quiet_fail = quiet_success = true;
		else if (strcmp(*argv, "quiet_fail") == 0)
			quiet_fail = true;
		else if (strcmp(*argv, "quiet_success") == 0)
			quiet_success = true;
		else if (strncmp(*argv, "recursion_limit=", 16) == 0)
			recursion_limit = strtoul(*argv + 16, NULL, 0);
		else
			break;
	}
	/* Process options.
	 */
	if (disable || enable) {
		int ret;
		char const *service = NULL;
		if ((ret = pam_get_item(
			pamh,
			PAM_SERVICE,
			(void const **)&service
			)) != PAM_SUCCESS)
			return ret;
		if (!service || !*service) {
			if (debug)
				pam_syslog(pamh, LOG_DEBUG, "no service");
			return PAM_IGNORE;
		}
		if (disable && in_list(disable, ':', service)) {
			if (debug)
				pam_syslog(
					pamh,
					LOG_DEBUG,
					"disabled"
					" for service %s"
					" due to disable=%s",
					service,
					disable
					);
			return PAM_IGNORE;
		}
		if (enable && !in_list(enable, ':', service)) {
			if (debug)
				pam_syslog(
					pamh,
					LOG_DEBUG,
					"not enabled"
					" for service %s"
					" due to enable=%s",
					service,
					enable
					);
			return PAM_IGNORE;
		}
	}
	/* Retrieve SSH authentication information.
	 */
	char const* ssh_auth_info = pam_getenv(pamh, "SSH_AUTH_INFO_0");
	if (!ssh_auth_info || !*ssh_auth_info) {
		if (debug)
			pam_syslog(
				pamh,
				LOG_DEBUG,
				!ssh_auth_info ? "no %s" : "empty %s",
				"SSH_AUTH_INFO_0"
				);
		return PAM_IGNORE;
	}
	/* Process SSH authentication information patterns.
	 */
	bool success = match_style != MATCH_ANY_OF;
	for (; argc > 0; --argc, ++argv) {
		bool matches = false;
		for (char const *s = ssh_auth_info; *s; s = next_line(s)) {
			bool const allow_prefix_match = true;
			matches = first_line_tokens_match(
				s,
				*argv,
				allow_prefix_match,
				recursion_limit
				);
			if (debug)
				pam_syslog(
					pamh,
					LOG_DEBUG,
					"ssh auth info"
					" line \"%.*s\""
					" %s"
					" pattern \"%s\"",
					(int)strcspn(s, "\n"),
					s,
					matches ? "matches" : "does not match",
					*argv
					);
			if (matches)
				break;
		}
		switch (match_style) {
		case MATCH_ALL_OF:
			if (matches)
				continue;
			success = false;
			break;
		case MATCH_ANY_OF:
			if (!matches)
				continue;
			success = true;
			break;
		case MATCH_NONE_OF:
			if (!matches)
				continue;
			success = false;
			break;
		}
		break;
	}
	if (!(success ? quiet_success : quiet_fail)) {
		char const *user = NULL;
		if (pam_get_item(
			pamh,
			PAM_USER,
			(void const **)&user
			) != PAM_SUCCESS || user == NULL)
			user = "(unknown)";
		pam_syslog(
			pamh,
			LOG_INFO,
			"ssh auth info %s%s%s%s %s by user %s",
			argc ? "pattern requirement" : "pattern requirements",
			argc ? " \"" : "",
			argc ? *argv : "",
			argc ? "\""  : "",
			success ? "met" : "not met",
			user
			);
	}
	return success ? PAM_SUCCESS : PAM_AUTH_ERR;
}

int
pam_sm_setcred(
	pam_handle_t *pamh,
	int flags,
	int argc,
	char const **argv
	) {
	return pam_sm_authenticate(pamh, flags, argc, argv);
}

int
pam_sm_acct_mgmt(
	pam_handle_t *pamh,
	int flags,
	int argc,
	char const **argv
	) {
	return pam_sm_authenticate(pamh, flags, argc, argv);
}

int
pam_sm_open_session(
	pam_handle_t *pamh,
	int flags,
	int argc,
	char const **argv
	) {
	return pam_sm_authenticate(pamh, flags, argc, argv);
}

int
pam_sm_close_session(
	pam_handle_t *pamh,
	int flags,
	int argc,
	char const **argv
	) {
	return pam_sm_authenticate(pamh, flags, argc, argv);
}

int
pam_sm_chauthtok(
	pam_handle_t *pamh,
	int flags,
	int argc,
	char const **argv
	) {
	return pam_sm_authenticate(pamh, flags, argc, argv);
}
